v2/pkg/engine/plan/interface_selection_rewriter.go

```diff
@@ -0,0 +1,359 @@
+		definition: definition,
+	}
+}
+
+func (r *interfaceSelectionRewriter) isFieldReturnsInterface(fieldRef int, enclosingNode ast.Node) (interfaceDefRef int, isInterface bool) {
+	var (
+		fieldDefRef int
+		hasField    bool
+	)
+
+	interfaceDefRef = ast.InvalidRef
+
+	switch enclosingNode.Kind {
+	case ast.NodeKindObjectTypeDefinition:
+		fieldDefRef, hasField = r.definition.ObjectTypeDefinitionFieldWithName(enclosingNode.Ref, r.operation.FieldNameBytes(fieldRef))
+		if !hasField {
+			return
+		}
+	case ast.NodeKindInterfaceTypeDefinition:
+		fieldDefRef, hasField = r.definition.InterfaceTypeDefinitionFieldWithName(enclosingNode.Ref, r.operation.FieldNameBytes(fieldRef))
+		if !hasField {
+			return
+		}
+	default:
+		return
+	}
+
+	fieldDefTypeName := r.definition.FieldDefinitionTypeNameBytes(fieldDefRef)
+	node, hasNode := r.definition.NodeByName(fieldDefTypeName)
+	if !hasNode {
+		return
+	}
+
+	if node.Kind != ast.NodeKindInterfaceTypeDefinition {
+		return
+	}
+
+	return node.Ref, true
+}
+
+func (r *interfaceSelectionRewriter) objectTypesImplementingInterface(interfaceDefRef int) (typeNames []string, ok bool) {
+	implementedByNodes := r.definition.InterfaceTypeDefinitionImplementedByRootNodes(interfaceDefRef)
+
+	typeNames = make([]string, 0, len(implementedByNodes))
+	for _, implementedByNode := range implementedByNodes {
+		if implementedByNode.Kind != ast.NodeKindObjectTypeDefinition {
+			continue
+		}
+
+		typeNames = append(typeNames, implementedByNode.NameString(r.definition))
+	}
+
+	if len(typeNames) > 0 {
+		return typeNames, true
+
+	}
+
+	return nil, false
+}
+
+func (r *interfaceSelectionRewriter) datasourceHasEntitiesWithName(dsConfiguration *DataSourceConfiguration, typeNames []string) (entityNames []string, ok bool) {
+	hasEntities := false
+	for _, typeName := range typeNames {
+		if len(dsConfiguration.RequiredFieldsByKey(typeName)) > 0 {
+			hasEntities = true
+			break
+		}
+	}
+
+	if !hasEntities {
+		return nil, false
+	}
+
+	entityNames = make([]string, 0, len(typeNames))
+	for _, typeName := range typeNames {
+		if dsConfiguration.HasRootNodeWithTypename(typeName) {
+			entityNames = append(entityNames, typeName)
+		}
+	}
+
+	return entityNames, true
+}
+
+type interfaceFieldSelectionInfo struct {
+	sharedFields            []sharedFieldSelection
+	inlineFragments         []inlineFragmentSelection
+	entitiesWithoutFragment []string
+}
+
+type sharedFieldSelection struct {
+	fieldSelectionRef int
+	fieldName         string
+}
+
+type inlineFragmentSelection struct {
+	selectionRef int
+	typeName     string
+	fieldNames   []string
+}
+
+func (r *interfaceSelectionRewriter) collectFieldInformation(fieldRef int, entityNames []string) interfaceFieldSelectionInfo {
+	fieldSelectionSetRef, ok := r.operation.FieldSelectionSet(fieldRef)
+	if !ok {
+		panic("unexpected error: field does not have a selection set")
+	}
+
+	fieldSelections := r.operation.SelectionSetFieldSelections(fieldSelectionSetRef)
```

Comment:
Does this recursively find all nested fields?

Reply:
no, only top level

we do not need to go deep